### PR TITLE
Uncouple config.Store from redis, and implement ConsulBackend

### PR DESCRIPTION
**Pull request by jbardin**

### CODE REVIEW

**Comment by jbardin** [comment]
Ah, yeah, didn't rewrite the MemoryBackend or tests ;(

**Comment by jbardin** [comment]
I'll rebase this after I get another commit in here. It looks like I need to refactor the hosts tracking sooner rather than later.

File: config/consul.go [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	ad := appCfg.(*AppDefinition)
+	for i := range ad.Assignments {
+		if ad.Assignments[i].Pool == pool {
+			// remove the item for the slice
+			a := ad.Assignments
+			a[i], a = a[len(a)-1], a[:len(a)-1]
```
Comment (joeshaw) [inline config/consul.go]: This doesn't look quite right to me.  If `a` were `{"foo", "bar", "baz", "quux"}` I think this results in `a` being `{"foo", "quux", "bar", "quux"}` -- the last element isn't sliced off.

Comment (joeshaw) [inline config/consul.go]: assuming i was looking for `"bar"` in the above example.

Comment (joeshaw) [inline config/consul.go]: Nope, never mind, I ran it through a playground and I grok it now.

**Comment by joeshaw** [comment]
:+1:

**Pull request merged**